Adds support for importing keys

yuce opened the pull request:
This PR works against Pilosa master
~This PR works against: https://github.com/pilosa/pilosa/pull/1557 There are 3 failing tests regarding exports with keys, but those tests should pass once https://github.com/pilosa/pilosa/issues/1580 is resolved.~

yuce commented:
@jaffee Is it OK to merge this one as well?

jaffee commented on client.go:
```diff
@@ -489,10 +500,21 @@ func (c *Client) httpRequest(method string, path string, data []byte, headers ma
+
+		if useCoordinator {
+			// TODO: we shouldn't need to fetch the coordinator every time
+			node, err := c.fetchCoordinatorNode()
```
I feel like we should only be fetching the coordinator node if a request fails because it should have been directed to the coordinator, but wasn't.

yuce replied:
I agree. I thought about adding a `coordinatorNode` field to the `PilosaClient` struct but then I have to add a mutex to prevent races. This is meant to be something that's safe until we find a better solution (like initializing `PilosaClient` with the coordinator node, etc.)

jaffee replied:
ok, I agree we can defer further work on this

jaffee commented on client.go:
```diff
@@ -642,40 +664,101 @@ func makeRequestData(query string, options *QueryOptions) ([]byte, error) {
+		columnIDs = make([]uint64, 0, recordCount)
+	}
+
+	if hasRowKeys && hasColKeys {
```
I'm thinking this would be more readable as a single `for` loop with two if/else blocks inside (similar to the above two for making the slices).

yuce replied:
I agree, it would be more readable. But I've unrolled the for loop there to make it faster.

yuce replied:
The non-scientific benchmark I've ran with the following script shows that unrolled version is about 1 seconds faster for a batch of 200 million on my machine but simplified this and its field values counterpart anyway.

yuce commented on imports.go:
```diff
@@ -87,24 +89,51 @@ func (b Column) Less(other Record) bool {
 
-func ColumnCSVUnmarshaller() CSVRecordUnmarshaller {
-	return ColumnCSVUnmarshallerWithTimestamp("")
+type CSVFormat uint
```
I agree again. Maybe we should remove csv stuff altogether (maybe move that kind of code to its own conveniences repo).

jaffee replied:
the pdk might be a good place - it already has some code for csv handling

yuce replied:
Moved the `csv` stuff to its package but then I had to move some stuff to the `imports` package, since `PilosaClient` depends on the csv stuff for exports (and vice versa). That change increased the lines to be reviewed quite a bit, but most of them are renames, so hopefully that would be trivial.

jaffee replied:
I'm a bit confused about the exporter - shouldn't it basically just return an io.Reader so that the user can write the exported data to wherever?

yuce replied:
I am not sure whether the user would like to write the exported data verbatim, they may transform it or maybe not write it all. Since we keep the CSV stuff in a subpackage, I think it makes sense to use it to make the exported data more useful for the user.

jaffee replied:
One could imagine Pilosa's export functionality being a lot more general than just dumping a CSV. It could support multiple content types - perhaps even in a pluggable way. I think that the responsibility of decoding that data falls outside the go-pilosa package. It also goes against the package structure that we've generally adopted to have the core package depend on subpackages - we've striven to make the reverse true in Pilosa.

yuce replied:
OK, ditched the imports package, so the core doesn't depend on it, but that requires refactoring the integration tests. Hold on...

jaffee replied:
sorry :/

yuce commented:
Refactoring the integration tests will take longer than I assumed. Have to
finish it later.

On Fri, Aug 24, 2018, 23:57 Matthew Jaffee <notifications@github.com> wrote:

> *@jaffee* commented on this pull request.
> ------------------------------
>
> In imports.go
> <https://github.com/pilosa/go-pilosa/pull/166#discussion_r212753380>:
>
> > @@ -87,24 +89,51 @@ func (b Column) Less(other Record) bool {
>  	return false
>  }
>
> -func ColumnCSVUnmarshaller() CSVRecordUnmarshaller {
> -	return ColumnCSVUnmarshallerWithTimestamp("")
> +type CSVFormat uint
>
> sorry :/
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pilosa/go-pilosa/pull/166#discussion_r212753380>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAfUWkqquA-Ef3Xkl_2uLhdQ6gnE2BE7ks5uUGjLgaJpZM4VuFIP>
> .
>

yuce commented:
@jaffee Removed the imports package, the PR is good to review again.

yuce commented:
ping!

jaffee commented on client.go:
```diff
@@ -489,17 +491,40 @@ func (c *Client) httpRequest(method string, path string, data []byte, headers ma
+			host = c.cluster.Host()
+			if host == nil {
+				return nil, nil, ErrEmptyCluster
+			}
```
I think the logic in this whole if/else block should be wrapped up in cluster.Host and Host should be modified to take a boolean `useCoordinator` argument.

yuce replied:
I've thought about moving that logic to the cluster (probably `cluster.Host`) before but that creates a chicken-egg problem: we can't get the coordinator node without creating the client first; and creating the client requires a cluster. That if/else code is messy though, I'll move it to a separate function.

jaffee commented on client.go:
```diff
@@ -489,17 +491,40 @@ func (c *Client) httpRequest(method string, path string, data []byte, headers ma
+					return nil, nil, errors.Wrap(err, "fetching coordinator node")
+				}
+				coordinatorURI = URIFromAddress(fmt.Sprintf("%s://%s:%d", node.Scheme, node.Host, node.Port))
+				c.coordinatorLock.Lock()
```
I *think* this lock should be acquired before calling `fetchCoordinatorNode` to avoid multiple goroutines trying to fetch the coordinator simultaneously. This also needs to do a final check of c.coordinatorURI after acquriring the write lock in case it was set by another goroutine between releasing the read lock and acquiring the write lock.

yuce replied:
I've locked the mutex after fetching the coordinator node to prevent other goroutines block reading the coordinatorURI, but I am not sure my logic in that code is correct. I'm adding a test to see whether anything blocks.

jaffee commented on client.go:
```diff
@@ -521,6 +527,33 @@ func (c *Client) httpRequest(method string, path string, data []byte, headers ma
+		coordinatorURI := c.coordinatorURI
+		c.coordinatorLock.RUnlock()
+		if coordinatorURI == nil {
+			node, err := c.fetchCoordinatorNode()
```
I believe that we need to obtain the write lock before calling `fetchCoordinatorNode`, my reasoning is as follows:

multiple routines may call Client.host simultaneously. They might simultaneously obtain a read lock and see that the coordinator is nil, they would then all try to fetch the coordinator node simultaneously which would be wasteful since only one needs to do it. Then they will all try to obtain the write lock to set the coordinator node to what they found, and each one will succeed in succession.

What we should do is obtain the write lock, then *re-check that c.coordinatorURI is still nil*, if it is still nil, then proceed with fetchCoordinator node. If it was set in between releasing the RLock and obtaining the Lock, then we can release the lock and use it.

yuce commented:
@jaffee I've updated the PR with your feedback and a test. I'll need to make the test more robust, other  than that it should be OK to review.

yuce commented:
@jaffee  I've reverted the coordinator lock logic, so it locks after fetching the coordinator node. Without that, `client.host` would sometimes return `nil` which causes a panic. I didn't get that problem once I've reverted the logic.

jaffee commented:
It was returning nil because there was a bug - if you unrevert it, I'll comment on where the change needs to be.

yuce commented:
unreverted

yuce commented on client.go:
```diff
@@ -521,6 +527,36 @@ func (c *Client) httpRequest(method string, path string, data []byte, headers ma
+		c.coordinatorLock.RUnlock()
+		if coordinatorURI == nil {
+			c.coordinatorLock.Lock()
+			if c.coordinatorURI == nil {
```
How did I miss that??? Thanks for pointing that out. Updated.

yuce commented:
I've simplified the coordinator lock logic a bit; removed the `coordinatorURI == nil` check after the lock. Rarely `coordinatorURI != nil` and the latest `else` we've added is not testable (since it occurs rarely).

jaffee commented:
now it's possible for many goroutines to get piled up on the write lock, and then each will re-fetch the coordinator node in series.

jaffee commented:
for what it's worth, I'm fine with that line not being covered reliably. As long as we have a test which calls `host` concurrently in several goroutines, gets the right answer, and passes the race detector, I'm satisfied that it's sufficiently covered.

yuce commented:
Yeah, but leaving that line drops the coverage. I want to try a few things before I give up.

jaffee commented:
I figured as much ;-) just letting you know my position

yuce commented:
@jaffee Improved the coverage.

yuce commented:
Added back to `nil` check too.
